ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureStorageCompactor.java

```diff
@@ -257,17 +289,67 @@ int purgeBlobs(List<CloudBlobMetadata> blobMetadataList) throws CloudStorageExce
   }
 
   /**
-   * Returns the dead blob in the specified partition with the earliest expiration or deletion time.
+   * Get the current compaction progress for a partition.
    * @param partitionPath the partition to check.
-   * @param fieldName the field name to use (expiration or deletion time).
-   * @return the {@link CloudBlobMetadata} for the dead blob, or NULL if none was found.
-   * @throws CloudStorageException
+   * @return a {@link Map} containing the progress time for compaction based on deletion and expiration time.
+   * @throws CloudStorageException if the operation fails.
+   */
+  Map<String, Long> getCompactionProgress(String partitionPath) throws CloudStorageException {
+    ByteArrayOutputStream baos = new ByteArrayOutputStream(CHECKPOINT_BUFFER_SIZE);
+    boolean hasCheckpoint = requestAgent.doWithRetries(
+        () -> azureBlobDataAccessor.downloadFile(CHECKPOINT_CONTAINER, partitionPath, baos, false),
```

Comment:
Using the same ByteArrayOutputStream for all retries could cause issues if the first try partially wrote to BAOS and then the second try will just start appending to the same stream.
Instead you could do something like this:
```java
    ByteArrayOutputStream baos = requestAgent.doWithRetries(() -> {
      ByteArrayOutputStream os = new ByteArrayOutputStream(CHECKPOINT_BUFFER_SIZE);
      boolean hasCheckpoint = azureBlobDataAccessor.downloadFile(CHECKPOINT_CONTAINER, partitionPath, os, false);
      return hasCheckpoint ? os : null;
    }, "Download compaction checkpoint", partitionPath);
```

Reply:
Good catch.  Moved the outputstream inside the lambda.